asciidtype/asciidtype/src/casts.c

```diff
@@ -86,6 +87,190 @@ ascii_to_ascii_get_loop(PyArrayMethod_Context *context, int aligned,
+{
+    PyArray_Descr **descrs = context->descriptors;
+    long in_size = (descrs[0]->elsize) / 4;
+    long out_size = ((ASCIIDTypeObject *)descrs[1])->size;
+    long copy_size;
+
+    if (out_size > in_size) {
+        copy_size = in_size;
+    }
+    else {
+        copy_size = out_size;
+    }
+
+    npy_intp N = dimensions[0];
+    char *in = data[0];
+    char *out = data[1];
+    npy_intp in_stride = strides[0];
+    npy_intp out_stride = strides[1];
+
+    while (N--) {
+        // copy input characters, checking that input UCS4
+        // characters are all ascii, raising an error otherwise
+        for (int i = 0; i < copy_size; i++) {
+            if (ucs4_character_is_ascii(in) == -1) {
+                PyGILState_STATE gstate;
+                gstate = PyGILState_Ensure();
+                PyErr_SetString(
+                        PyExc_TypeError,
+                        "Can only store ASCII text in a ASCIIDType array.");
+                PyGILState_Release(gstate);
+                return -1;
+            }
+            // UCS4 character is ascii, so copy first byte of character
+            // into output, ignoring the rest
+            *(out + i) = *(in + i * 4);
+        }
+        // write zeros to remaining ASCII characters (if any)
+        for (int i = copy_size; i < out_size; i++) {
+            *(out + i) = '\0';
+        }
+        in += in_stride;
+        out += out_stride;
+    }
+
+    return 0;
+}
+
+static int
+unicode_to_ascii_get_loop(PyArrayMethod_Context *NPY_UNUSED(context),
+                          int NPY_UNUSED(aligned),
+                          int NPY_UNUSED(move_references),
+                          const npy_intp *NPY_UNUSED(strides),
+                          PyArrayMethod_StridedLoop **out_loop,
+                          NpyAuxData **NPY_UNUSED(out_transferdata),
+                          NPY_ARRAYMETHOD_FLAGS *flags)
+{
+    *out_loop = (PyArrayMethod_StridedLoop *)&unicode_to_ascii;
+
+    *flags = 0;
+    return 0;
+}
+
+static int
+ascii_to_unicode(PyArrayMethod_Context *context, char *const data[],
+                 npy_intp const dimensions[], npy_intp const strides[],
+                 NpyAuxData *NPY_UNUSED(auxdata))
+{
+    PyArray_Descr **descrs = context->descriptors;
+    long in_size = ((ASCIIDTypeObject *)descrs[0])->size;
+    long out_size = (descrs[1]->elsize) / 4;
+    long copy_size;
+
+    if (out_size > in_size) {
+        copy_size = in_size;
+    }
+    else {
+        copy_size = out_size;
+    }
+
+    npy_intp N = dimensions[0];
+    char *in = data[0];
+    char *out = data[1];
+    npy_intp in_stride = strides[0];
+    npy_intp out_stride = strides[1];
+
+    while (N--) {
+        // copy ASCII input to first byte, fill rest with zeros
+        for (int i = 0; i < copy_size; i++) {
+            *(out + i * 4) = *(in + i);
+            for (int j = 1; j < 4; j++) {
+                *(out + i * 4 + j) = '\0';
+            }
```

Comment:
If you don't bother about supporting unaligned access, you could just write:
```
((Py_UCS4 *)out)[i] = in[i];
```

Reply:
Ah, you need to make sure to use `unsigned char` in that case.  But that would also make the check for ASCII much more clear.  (Could also use `Py_UCS1` if you like.)